pandas/_libs/src/klib/khash_python.h

```diff
@@ -330,6 +333,40 @@ Py_hash_t PANDAS_INLINE tupleobject_hash(PyTupleObject* key) {
 }
 
 
+// TODO: same thing for timedelta64 objects
+Py_hash_t PANDAS_INLINE np_datetime64_object_hash(PyObject* key) {
+    // GH#50690 numpy's hash implementation does not preserve comparabity
+    // either across resolutions or with standard library objects.
+    NPY_DATETIMEUNIT unit = (NPY_DATETIMEUNIT)((PyDatetimeScalarObject*)key)->obmeta.base;
+    npy_datetime value = ((PyDatetimeScalarObject*)key)->obval;
+    npy_datetimestruct dts;
+    PyObject* dt;
+
+    pandas_datetime_to_datetimestruct(value, unit, &dts);
+
+    if ((dts.year > 0) && (dts.year <= 9999) && (dts.ps == 0) && (dts.as == 0)) {
+        // we CAN cast to pydatetime, so use that hash to ensure we compare
+        // as matching standard library datetimes (and pd.Timestamps)
+        PyDateTime_IMPORT;
```

Comment:
Is there a way to move this out of the function body? Maybe it can go into the global hashtable code instead?

Reply:
that causes a complaint at build time.  there is another use of PyDateTime_IMPORT in the json code that is also runtime

Reply:
So does this not work to move to the global hashtable code in the cython file? Also agree with @seberg would like to avoid using the macro here

Reply:
ill defer to you guys on this, but im uncomfortable with having this c file have an implicit dependency on the higher-level cython file

Reply:
note there's also a coment in np_datetime.c saying it would be helpful to get PyDateTime_IMPORT to work in that file.

Reply:
Yes, the implicit dependency is not nice, but I think that is just something to live with.  You need some initialization, which needs to be a function call and that cannot happen in a header file.  I would just say: If you forget, you get a clean segfault pointing at the first use of Datetime code in gdb/lldb.  We put a comment there explaining things.

Reply:
Just some (hopefully helpful) context - the header file never gets "executed" in any sense, so just putting the macro in a header file doesn't by itself just run the code you've inserted. 

The entry point for C extensions is a module initialization function like you see in the Python documentation - this gets executed when the Python interpreter loads the extension:

https://docs.python.org/3/extending/extending.html#the-module-s-method-table-and-initialization-function

I am assuming that the global namespace of the Cython file is akin to the module initialization; if that is true then it would be natural for the import to go there rather than the header.

With that said the workaround suggested here is fine, but I don't think the dependency problem you were worried about is too big of a deal